### DISCUSSION TITLE
Hanging issues

### DISCUSSION THREAD

**Discussion by chanon** [porsager/postgres#967] [General]
EDIT: I initially wrote this in a more aggressive tone, so toned it down.

This package's performance is very good and I like the interface a lot. But there is an alarming category of issues where it seems a deadlock happens. For many of these issues the behavior is that once it occurs, the whole app will not be able to do any additional database requests because async calls to the library will never return again.

Examples include:
- #751, #925 
- #827 
- #893 
- #941 
- #966 

Some of the above have very clear reproductions and when using the library it is possible to work around when you know they cause hangs. However it doesn't inspire much confidence when maybe an unknown case could cause your whole service to stop responding to requests. Users in #827 for example haven't seem to figured out why their services hanged.

I just wanted to bring attention to these issues, and am wondering if there might be a common root cause?

**Comment by porsager** [comment]
Thanks for the thoughts, and yes, I haven't had time to look into any of these things, as they don't affect any of my own projects where I use my library. I'm using this in many various high load production setups and haven't seen any of these issues myself. It's not to say the issues aren't real or that they aren't faults in this library, they probably are, but just to let you know why they haven't gotten any focus from my end.

I usually manage to find a couple of days time to focus now and then, but I've been on a long busy stretch, so it's been a while since I did that. Now I've previously had some users pay me to look at bugs which allows me to prioritize them, so I'd say any serious production usage is still safe.

**Reply by chanon** [reply]
Good to hear about your stable usages in high load production setups. I will keep monitoring my use to see if there are any other similar cases. Thank you for your amazing work again.

I'm thinking there might be a single (simple?) root cause for these as the behavior / what happens is very similar.
